src/createApp.ts

```diff
@@ -1044,9 +1044,20 @@ const createApp = compose({
 			});
 		},
 
-		hasWidget(this: App, id: Identifier): boolean {
+		hasWidget(this: App, id: Identifier): Promise<boolean> {
+			const { defaultWidgetStore } = this;
+
+			let exists: Promise<boolean> | boolean = widgetFactories.get(this).has(id) || widgetInstances.get(this).has(id);
+
+			if (!exists && defaultWidgetStore) {
+				const registeredCustomElementFactories = customElementFactories.get(this);
+				exists = defaultWidgetStore.get(id).then((state: any) => {
+					return registeredCustomElementFactories.has(state.type);
+				});
+			}
+
 			// See if there is a widget factory, or else an existing custom element instance.
-			return widgetFactories.get(this).has(id) || widgetInstances.get(this).has(id);
+			return Promise.resolve(exists);
```

Comment:
Could return early with `Promise.resolve(exists)` if `exists || !defaultWidgetStore`, and return the promise for `registeredCustomElementFactories.has()` otherwise. Saves an unnecessary promise allocation.